libs/SmartSync/SmartSync/Classes/Manager/SFSmartSyncSyncManager.m

```diff
@@ -275,46 +277,51 @@ - (void) syncDown:(SFSyncState*)sync updateSync:(SyncUpdateBlock)updateSync fail
 
     __block NSUInteger countFetched = 0;
     __block NSUInteger totalSize = 0;
-    __block SFSyncDownTargetFetchCompleteBlock completeBlockRecurse = ^(NSArray *records) {};
+    __block NSUInteger progress = 0;
+    __block SFSyncDownTargetFetchCompleteBlock continueFetchBlockRecurse = ^(NSArray *records) {};
     __weak SFSmartSyncSyncManager *weakSelf = self;
     
-    SFSyncDownTargetFetchCompleteBlock completeBlock = ^(NSArray* records) {
+    SFSyncDownTargetFetchCompleteBlock startFetchBlock = ^(NSArray* records) {
         totalSize = target.totalSize;
-        if (countFetched == 0) { // after first request only
-            updateSync(nil, totalSize == 0 ? 100 : 0, totalSize, kSyncManagerUnchanged);
-            if (totalSize == 0) {
-                return;
+        updateSync(nil, totalSize == 0 ? 100 : 0, totalSize, kSyncManagerUnchanged);
+        if (totalSize != 0) continueFetchBlockRecurse(records);
+    };
+
+    SFSyncDownTargetFetchCompleteBlock continueFetchBlock = ^(NSArray* records) {
+        __strong SFSmartSyncSyncManager *strongSelf = weakSelf;
+        if (records != nil) {
```

Comment:
The sync down target is supposed to return nil when down - it might return empty arrays before (without being done) - for instance for a refresh-sync-down, we do soql calls with ids we found locally, the ids might be gone from the server and therefore some of the soql calls could possibly return no records.